lockfile_test.go

```diff
@@ -0,0 +1,553 @@
+	defer os.Remove(l.name)
+
+	l.Lock()
+	assert.True(t, l.Locked(), "Locked() said we didn't have a write lock")
+	l.Unlock()
+}
+
+func TestROLockfileWrite(t *testing.T) {
+	l, err := getTempROLockfile()
+	require.Nil(t, err, "error getting temporary lock file")
+	defer os.Remove(l.name)
+
+	l.Lock()
+	assert.False(t, l.Locked(), "Locked() said we have a write lock")
+	l.Unlock()
+}
+
+func TestLockfileTouch(t *testing.T) {
+	l, err := getTempLockfile()
+	require.Nil(t, err, "error getting temporary lock file")
+	defer os.Remove(l.name)
+
+	l.Lock()
+	m, err := l.Modified()
+	require.Nil(t, err, "got an error from Modified()")
+	assert.True(t, m, "new lock file does not appear to have changed")
+
+	now := time.Now()
+	assert.False(t, l.TouchedSince(now), "file timestamp was updated for no reason")
+
+	time.Sleep(2 * time.Second)
+	err = l.Touch()
+	require.Nil(t, err, "got an error from Touch()")
+	assert.True(t, l.TouchedSince(now), "file timestamp was not updated by Touch()")
+
+	m, err = l.Modified()
+	require.Nil(t, err, "got an error from Modified()")
+	assert.False(t, m, "lock file mistakenly indicated that someone else has modified it")
+
+	stdin, stdout, stderr, err := subTouch(l)
+	require.Nil(t, err, "got an error starting a subprocess to touch the lockfile")
+	l.Unlock()
+	io.Copy(ioutil.Discard, stdout)
+	stdin.Close()
+	io.Copy(ioutil.Discard, stderr)
+	l.Lock()
+	m, err = l.Modified()
+	l.Unlock()
+	require.Nil(t, err, "got an error from Modified()")
+	assert.True(t, m, "lock file failed to notice that someone else modified it")
+}
+
+func TestLockfileWriteConcurrent(t *testing.T) {
+	l, err := getTempLockfile()
+	require.Nil(t, err, "error getting temporary lock file")
+	defer os.Remove(l.name)
+	var wg sync.WaitGroup
+	var highestMutex sync.Mutex
+	var counter, highest int64
+	for i := 0; i < 100000; i++ {
+		wg.Add(1)
+		go func() {
+			l.Lock()
+			tmp := atomic.AddInt64(&counter, 1)
+			assert.True(t, tmp >= 0, "counter should never be less than zero")
+			highestMutex.Lock()
+			if tmp > highest {
+				// multiple writers should not be able to hold
+				// this lock at the same time, so there should
+				// be no point at which two goroutines are
+				// between the AddInt64() above and the one
+				// below
+				highest = tmp
+			}
+			highestMutex.Unlock()
+			atomic.AddInt64(&counter, -1)
+			l.Unlock()
+			wg.Done()
+		}()
+	}
+	wg.Wait()
+	assert.True(t, highest == 1, "counter should never have gone above 1, got to %d", highest)
+}
+
+func TestLockfileReadConcurrent(t *testing.T) {
+	l, err := getTempLockfile()
+	require.Nil(t, err, "error getting temporary lock file")
+	defer os.Remove(l.name)
+	var wg sync.WaitGroup
+	var counter, highest int64
+	var highestMutex sync.Mutex
+	for i := 0; i < 100000; i++ {
+		wg.Add(1)
+		go func() {
+			l.RLock()
+			tmp := atomic.AddInt64(&counter, 1)
+			assert.True(t, tmp >= 0, "counter should never be less than zero")
+			highestMutex.Lock()
```

Comment:
I think we could do a `time.Sleep(500 * time.Milisecond)` after the mutex acquisition and reduce the loop range to 10? It would stress the machines a bit less while remaining "quite confident" we're forcing a race. Same could be done for the writer test, and maybe the others as well.